modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPreloader.java

```diff
@@ -228,8 +234,18 @@ public boolean disableRebalancingCancellationOptimization() {
                 if (grp.persistenceEnabled() && exchFut != null) {
                     List<UUID> nodeIds = exchFut.partitionHistorySupplier(grp.groupId(), p, part.initialUpdateCounter());
 
-                    if (!F.isEmpty(nodeIds))
+                    if (!F.isEmpty(nodeIds)) {
+                        if (sameDcNodeIds != null) {
```

Comment:
I suggest to reuse this code for the very similar block on the line 279: just unify filtering operation (it is possible to use `removeIf` here instead of `retainAll`) and move it to a separate method with a name like `filterNodesFromRemoteDcs`.

It could look like this:
`
    private <T> List<T> filterNodesFromRemoteDcs(List<T> nodesToFilter, Collection<UUID> nodesToRetain, Predicate<T> filter) {
        if (nodesToRetain == null)
            return nodesToFilter;
        
        List<T> nodes0 = new ArrayList<>(nodesToFilter);

        nodes0.removeIf(filter);

        return !F.isEmpty(nodes0) ? nodes0 : nodesToFilter;
    }
`